src/main/java/com/timgroup/statsd/UnixStreamClientChannel.java

```diff
@@ -0,0 +1,127 @@
+package com.timgroup.statsd;
+
+import jnr.unixsocket.UnixSocketAddress;
+import jnr.unixsocket.UnixSocketChannel;
+import jnr.unixsocket.UnixSocketOptions;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.SocketChannel;
+
+/**
+ * A ClientChannel for Unix domain sockets.
+ */
+public class UnixStreamClientChannel implements ClientChannel {
+    private final UnixSocketAddress address;
+    private final int timeout;
+    private final int bufferSize;
+
+    private SocketChannel delegate;
+    private final ByteBuffer delimiterBuffer = ByteBuffer.allocateDirect(Integer.SIZE / Byte.SIZE).order(ByteOrder.LITTLE_ENDIAN);
+
+    /**
+     * Creates a new NamedPipeClientChannel with the given address.
+     *
+     * @param address Location of named pipe
+     */
+    UnixStreamClientChannel(SocketAddress address, int timeout, int bufferSize) throws IOException {
+        this.delegate = null;
+        this.address = (UnixSocketAddress) address;
+        this.timeout = timeout;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isConnected();
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        if (delegate == null || !delegate.isConnected()) {
+            connect();
+        }
+
+        int size = src.remaining();
+        if (size == 0) {
+            return 0;
+        }
+        delimiterBuffer.clear();
+        delimiterBuffer.putInt(size);
+        delimiterBuffer.flip();
+
+        try {
+            if (writeAll(delimiterBuffer, true) > 0) {
+                writeAll(src, false);
+            }
+        } catch (IOException e) {
+            delegate.close();
+            delegate = null;
+            throw e;
+        }
+
+        return size;
+    }
+
+    /**
+     * Writes all bytes from the given buffer to the channel.
+     * @param bb buffer to write
+     * @param canReturnOnTimeout if true, we return if the channel is blocking and we haven't written anything yet
+     * @return number of bytes written
+     * @throws IOException if the channel is closed or an error occurs
+     */
+    public int writeAll(ByteBuffer bb, boolean canReturnOnTimeout) throws IOException {
+        int remaining = bb.remaining();
+        int written = 0;
+        while (remaining > 0) {
+            int read = delegate.write(bb);
+
+            // If we haven't written anything yet, we can still return
+            if (read == 0 && canReturnOnTimeout && written == 0) {
```

Comment:
This should not get stuck in an endless loop even if we did a partial write.

Reply:
This should be better now